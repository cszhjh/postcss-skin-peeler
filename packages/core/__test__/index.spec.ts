import { dirname, relative, resolve } from 'node:path'
import postcss from 'postcss'
import { describe, expect, it } from 'vitest'
import plugin from '../src/main'
import type { PluginOptions } from '../src/types'

const FROM = resolve('__test__/style/index.css')
const IMG_SRC = '__test__/images/origin'
const GENERATE_SKIN_IMG = '__test__/images/generate'
const REPLACE_SKIN_IMG = '__test__/images/replace'
const ORIGIN_IMG_ONE = resolve(IMG_SRC, './one.jpg')
const ORIGIN_IMG_TWO = resolve(IMG_SRC, './two.jpg')
const ORIGIN_IMG_ONE_RELATIVE = relative(dirname(FROM), ORIGIN_IMG_ONE)
const ORIGIN_IMG_TWO_RELATIVE = relative(dirname(FROM), ORIGIN_IMG_TWO)

const PLUGIN_OPTIONS: PluginOptions[] = [
  {
    mode: 'generate',
    imgSrc: IMG_SRC,
    skinSrc: GENERATE_SKIN_IMG,
    prefixSelector: '.skin',
    coverSize: true,
  },
  {
    mode: 'replace',
    imgSrc: IMG_SRC,
    skinSrc: REPLACE_SKIN_IMG,
    coverSize: true,
  },
]

async function run(css: string, options: PluginOptions | PluginOptions[] = PLUGIN_OPTIONS) {
  const { css: output } = await postcss([plugin(options)]).process(css, { from: FROM })
  return output
}

describe('test url quotation mark', () => {
  it('test url without quotes', async () => {
    const input = `
      .non-quotation {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test url with single quotes', async () => {
    const input = `
      .single-quote {
        background: url('${ORIGIN_IMG_ONE_RELATIVE}') no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test url with double quotes', async () => {
    const input = `
      .double-quote {
        background: url("${ORIGIN_IMG_ONE_RELATIVE}") no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test url with protocol', () => {
  it('test url with http', async () => {
    const input = `
      .http {
        background: url(http://varletjs.org/varlet_icon.png) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toBe(input)
  })

  it('test url with https', async () => {
    const input = `
      .http {
        background: url(https://varletjs.org/varlet_icon.png) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toBe(input)
  })
})

describe('test background stack', () => {
  it('test background is stacked in the single rule', async () => {
    const input = `
      .rule-single-stacked {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
        background-image: url(${ORIGIN_IMG_TWO_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test background is stacked in the multiple rule', async () => {
    const input = `
      .rule-multiple-stacked {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }

      .rule-multiple-stacked {
        background: url(${ORIGIN_IMG_TWO_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test html and body combination', () => {
  it('test html as prefix selector', async () => {
    const input = `
      html .html-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test body as prefix selector', async () => {
    const input = `
      body .body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test html body as prefix selector', async () => {
    const input = `
      html body .html-body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test html follow selector as prefix selector', async () => {
    const input = `
      html.follow .html-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test body follow selector as prefix selector', async () => {
    const input = `
      body.follow .body-prefix {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test rule in @media', () => {
  const input = `
    @media screen and (min-width: 1024px) {
      .media {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    }
  `
  it('test rule in @media', async () => {
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })
})

describe('test cover size', () => {
  it('test default unit cover size', async () => {
    const input = `
      .wrapper {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const output = await run(input)
    expect(output).toMatchSnapshot()
  })

  it('test custom unit cover size', async () => {
    const input = `
      .wrapper {
        background: url(${ORIGIN_IMG_ONE_RELATIVE}) no-repeat;
      }
    `
    const coverSize = ({ width, height }: { width: number; height: number }) => {
      return { width: `${width / 100}rem`, height: `${height / 100}rem` }
    }

    const output = await run(input, [
      {
        mode: 'generate',
        imgSrc: IMG_SRC,
        skinSrc: GENERATE_SKIN_IMG,
        prefixSelector: '.skin',
        coverSize,
      },
      {
        mode: 'replace',
        imgSrc: IMG_SRC,
        skinSrc: REPLACE_SKIN_IMG,
        coverSize,
      },
    ])
    expect(output).toMatchSnapshot()
  })
})
