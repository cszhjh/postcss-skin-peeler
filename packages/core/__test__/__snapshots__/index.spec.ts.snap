// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test background stack > test background is stacked in the multiple rule 1`] = `
"
      .rule-multiple-stacked {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .rule-multiple-stacked {background-image: /* __DEV_GENERATE_REWRITE__ */ /* __DEV_GENERATE__ */ url("../images/generate/two.jpg");width: /* __DEV_GENERATE_REWRITE__ */ 280px;height: /* __DEV_GENERATE_REWRITE__ */ 350px;
      }

      .rule-multiple-stacked {
        background: /* __DEV_REPLACE__ */ url("../images/replace/two.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 350px;
        height: /* __DEV_REPLACE__ */ 350px;
      }
    "
`;

exports[`test background stack > test background is stacked in the single rule 1`] = `
"
      .rule-single-stacked {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        background-image: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .rule-single-stacked {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test cover size > test custom unit cover size 1`] = `
"
      .wrapper {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 3rem;
        height: /* __DEV_REPLACE__ */ 3rem;
      }
      .skin .wrapper {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 3rem;height: /* __DEV_GENERATE__ */ 2rem;
      }
    "
`;

exports[`test cover size > test default unit cover size 1`] = `
"
      .wrapper {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .wrapper {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test html and body combination > test body as prefix selector 1`] = `
"
      body .body-prefix {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      body .skin .body-prefix {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test html and body combination > test body follow selector as prefix selector 1`] = `
"
      body.follow .body-prefix {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      body.follow .skin .body-prefix {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test html and body combination > test html as prefix selector 1`] = `
"
      html .html-prefix {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      html .skin .html-prefix {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test html and body combination > test html body as prefix selector 1`] = `
"
      html body .html-body-prefix {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      html body .skin .html-body-prefix {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test html and body combination > test html follow selector as prefix selector 1`] = `
"
      html.follow .html-prefix {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      html.follow .skin .html-prefix {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test rule in @media > test rule in @media 1`] = `
"
    @media screen and (min-width: 1024px) {
      .media {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .media {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    }
  "
`;

exports[`test url quotation mark > test url with double quotes 1`] = `
"
      .double-quote {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .double-quote {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test url quotation mark > test url with single quotes 1`] = `
"
      .single-quote {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .single-quote {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;

exports[`test url quotation mark > test url without quotes 1`] = `
"
      .non-quotation {
        background: /* __DEV_REPLACE__ */ url("../images/replace/one.jpg") no-repeat;
        width: /* __DEV_REPLACE__ */ 300px;
        height: /* __DEV_REPLACE__ */ 300px;
      }
      .skin .non-quotation {background-image: /* __DEV_GENERATE__ */ url("../images/generate/one.jpg");width: /* __DEV_GENERATE__ */ 300px;height: /* __DEV_GENERATE__ */ 200px;
      }
    "
`;
